src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -368,78 +379,70 @@ public synchronized void removeShard(int shardId, String reason) throws Elastics
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close translog service", e, shardId);
-                // ignore
-            }
+            // now we can close the translog service, we need to close it before the we close the shard
+            close(sId, shardInjector, TranslogService.class);
             // this logic is tricky, we want to close the engine so we rollback the changes done to it
             // and close the shard so no operations are allowed to it
             if (indexShard != null) {
                 try {
-                    ((IndexShard) indexShard).close(reason);
+                    indexShard.close(reason);
                 } catch (Throwable e) {
                     logger.debug("[{}] failed to close index shard", e, shardId);
                     // ignore
                 }
             }
+            close(sId, shardInjector,
+                    Engine.class,
+                    MergeSchedulerProvider.class,
+                    MergePolicyProvider.class,
+                    IndexShardGatewayService.class,
+                    Translog.class,
+                    PercolatorQueriesRegistry.class);
+
+            // call this before we close the store, so we can release resources for it
+            indicesLifecycle.afterIndexShardClosed(sId, indexShard);
+        } finally {
             try {
-                shardInjector.getInstance(Engine.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close engine", e, shardId);
-                // ignore
-            }
-            try {
-                shardInjector.getInstance(MergeSchedulerProvider.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close merge policy scheduler", e, shardId);
-                // ignore
-            }
-            try {
-                shardInjector.getInstance(MergePolicyProvider.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close merge policy provider", e, shardId);
-                // ignore
-            }
-            try {
-                shardInjector.getInstance(IndexShardGatewayService.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close index shard gateway", e, shardId);
-                // ignore
-            }
-            try {
-                // now we can close the translog
-                shardInjector.getInstance(Translog.class).close();
+                shardInjector.getInstance(Store.class).close();
             } catch (Throwable e) {
-                logger.debug("[{}] failed to close translog", e, shardId);
-                // ignore
+                logger.warn("[{}] failed to close store on shard deletion", e, shardId);
             }
+        }
+    }
+
+    private void close(ShardId shardId, Injector shardInjector, Class<? extends Closeable>... toClose) {
```

Comment:
Just to avoid name collision (since there's already a public `close` method), can we rename this `closeWithInjector` or `innerClose` or something like that?